### DISCUSSION TITLE
How to interpret the results of LAP with GAE

### DISCUSSION THREAD

**Discussion by JackyYangPassion** [alibaba/GraphScope#2731] [General]
**Describe the bug**
DataSet
Vertices
```
id
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
```
Edges

```
from,to
2,6
0,16
0,11
0,4
3,13
0,7
0,14
0,10
1,17
2,5
3,9
2,8
9,15
2,12
```

when run lpa
```
# 查看算法结果
# run builtin applications
result_lpa = graphscope.lpa(sub_graph_with_path)
result_lpa.to_dataframe({"node": "v.id", "result": "r"})
```

the result:
![image](https://user-images.githubusercontent.com/13795366/231680118-10f8c811-54b7-44f7-b48a-2ae04945d376.png)

Use networkX 
Original Graph before clustering.
![image](https://user-images.githubusercontent.com/13795366/231680849-4136c654-4baf-4581-b3a6-e4c0f831c069.png)

Results after clustering.
![image](https://user-images.githubusercontent.com/13795366/231680910-9308309b-b2fc-4c23-bb71-6e6a69f63199.png)

**Expected behavior**
I want to know how can get the result like networkX with graphScope lpa

for example
community 0 [node_list]
community 1 [node_list]
.....
community n [node_list]

**Environment (please complete the following information):**
 - GraphScope version: 0.20.0
 - OS: linux

**Comment by siyuan0322** [comment]
Since the result of lpa is a pandas DataFrame, you could make some manipulations over it. For example, extract the nodes that related to a certain community like this:
```python
communities = {}
for row in df.iterrows():
    node_id, community_id = row[1]
    if community_id not in communities:
        communities[community_id] = []
    communities[community_id].append(node_id)

for k, v in communities.items():
    print('community {}: {}'.format(k, v))
```

**Comment by siyuan0322** [comment]
By the way, it's pretty suitable to use chatGPT-like tools to do these jobs. 🤣

**Comment by JackyYangPassion** [comment]
Yes, I consulted with Chat-GPT while writing the issue!

![image](https://user-images.githubusercontent.com/13795366/231702096-a5f31987-73e9-428c-b20f-1dbf8a45bc16.png)

Although there were some errors, it ran after the modifications were made.

**Comment by JackyYangPassion** [comment]
the result
![image](https://user-images.githubusercontent.com/13795366/231705400-32a88b0b-49dc-46a9-9dd2-2fd07341df52.png)

**Comment by siyuan0322** [comment]
It seems the result is not correct, according to the edge lists, do you mean this?

**Comment by JackyYangPassion** [comment]
Yes!
First  
 ```
 [0,4,7,10,11,14,16] should in the same community
  but [10],[11],[14] belong to different community
 ```
   
Second 
  ```
    community_id should Increment from 0
    community 0 []
    community 1 []
    ......
  ```

**Comment by siyuan0322** [comment]
Thanks, we will try to confirm and debug that. Sorry for the inconvenience.

**Comment by JackyYangPassion** [comment]
From the discussion #1498  cdlp = lpa in graphscope
What's the difference between networkX lpa and graphscope lpa?

**Comment by siyuan0322** [comment]
graphscope lpa is the version that we implement for LDBC benchmark, which optimized for speed.
networkx lpa is trying to be compatible with networkx native lpa.

**Comment by siyuan0322** [comment]
@JackyYangPassion 经过调查，发现原因
1. 我们这里的 CDLP 是根据 LDBC 的要求实现，实现细节可以参考 [这里](https://arxiv.org/pdf/2011.15028.pdf) 的 CDLP 一节。我在这大概讲一下过程：
2. 第一轮0号点的label初始化为0，4初始化为4；第二轮0的所有邻居都把label更新为0，但0号点的邻居里，label最小的是4（不和自身的label比较），所以它把自己的label更新为4，再之后，0的邻居实际上每轮都用0号点的label更新自己（不需要比较大小，只有0一个邻居），所以刚好也是0和4交替，只不过跟0号点相反。其实就是 label互相抖动，一个最小复现的case 就是两个点直接相连的情况。

可以配置 `cdlp(g,directed=False, max_round=99)` `cdlp(g,directed=False, max_round=100)` 来看一下区别。其中 directed 指定无向图，max_round 指定最大轮数。

**Comment by JackyYangPassion** [comment]
感谢，这个修复起来容易吗？可以指导下修复

**Comment by siyuan0322** [comment]
我看一下 GS 有没有其他实现

**Comment by siyuan0322** [comment]
研究了一些 networkx 的这个lpa 的版本，发现实现上有一些区别，比如这个 [asyn_lpa_communities](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.community.label_propagation.asyn_lpa_communities.html) 是不确定性的。我理解震荡会在一些特殊的节点内发生，这时由于每一轮迭代双方节点互换，所以达不到一个稳定状态。而且分布式比单机算法更加复杂一些，不容易收敛。这个如果数据量级大了之后应该不是什么问题。
另外，如果是找community 按您目前的需求感觉 cc 算法更合适一些，这个能给你想要的结果。

**Comment by JackyYangPassion** [comment]
好的 感谢建议，我试试

**Comment by JackyYangPassion** [comment]
> 研究了一些 networkx 的这个lpa 的版本，发现实现上有一些区别，比如这个 [asyn_lpa_communities](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.community.label_propagation.asyn_lpa_communities.html) 是不确定性的。我理解震荡会在一些特殊的节点内发生，这时由于每一轮迭代双方节点互换，所以达不到一个稳定状态。而且分布式比单机算法更加复杂一些，不容易收敛。这个如果数据量级大了之后应该不是什么问题。 另外，如果是找community 按您目前的需求感觉 cc 算法更合适一些，这个能给你想要的结果。

“cc 算法” 是指 wcc?

**Comment by siyuan0322** [comment]
> “cc 算法” 是指 wcc?

是的

**Comment by JackyYangPassion** [comment]
Mark After this period of study in the field of graph mining, I have a better understanding.
